lib/core/subrequest.c

```diff
@@ -0,0 +1,145 @@
+    }
+
+DEFINE_LOGGER(ssl, protocol_version)
+DEFINE_LOGGER(ssl, session_reused)
+DEFINE_LOGGER(ssl, cipher)
+DEFINE_LOGGER(ssl, cipher_bits)
+DEFINE_LOGGER(ssl, session_id)
+DEFINE_LOGGER(http1, request_index)
+DEFINE_LOGGER(http2, stream_id)
+DEFINE_LOGGER(http2, priority_received)
+DEFINE_LOGGER(http2, priority_received_exclusive)
+DEFINE_LOGGER(http2, priority_received_parent)
+DEFINE_LOGGER(http2, priority_received_weight)
+DEFINE_LOGGER(http2, priority_actual)
+DEFINE_LOGGER(http2, priority_actual_parent)
+DEFINE_LOGGER(http2, priority_actual_weight)
+
+#undef DEFINE_LOGGER
+
+static const h2o_conn_callbacks_t http1_callbacks = {
+    get_sockname, /* stringify address */
+    get_peername, /* ditto */
+    NULL,         /* push */
+    get_socket,   /* get underlying socket */
+    NULL,         /* get debug state */
+    {{
+        {log_protocol_version, log_session_reused, log_cipher, log_cipher_bits, log_session_id}, /* ssl */
+        {log_request_index},                                                                     /* http1 */
+        {NULL}                                                                                   /* http2 */
+    }}};
+
+static const h2o_conn_callbacks_t http2_callbacks = {
+    get_sockname,    /* stringify address */
+    get_peername,    /* ditto */
+    NULL,            /* push */
+    get_socket,      /* get underlying socket */
+    get_debug_state, /* get debug state */
+    {{
+        {log_protocol_version, log_session_reused, log_cipher, log_cipher_bits, log_session_id}, /* ssl */
+        {NULL},                                                                                  /* http1 */
+        {log_stream_id, log_priority_received, log_priority_received_exclusive, log_priority_received_parent,
+         log_priority_received_weight, log_priority_actual, log_priority_actual_parent, log_priority_actual_weight} /* http2 */
+    }}};
+
+static const h2o_conn_callbacks_t *get_subreq_callbacks(h2o_req_t *req)
+{
+    return req->version < 0x200 ? &http1_callbacks : &http2_callbacks;
+}
+
+h2o_req_t *h2o_create_subrequest(h2o_req_t *src)
+{
+    struct st_h2o_subreq_t *subreq = h2o_mem_alloc(sizeof(*subreq));
+    subreq->conn.ctx = src->conn->ctx;
+    subreq->conn.hosts = src->conn->hosts;
+    subreq->conn.connected_at = src->conn->connected_at;
+    subreq->conn.id = src->conn->id;
+    subreq->conn.callbacks = get_subreq_callbacks(src);
+    h2o_init_request(&subreq->super, &subreq->conn, NULL);
+
+    subreq->src = src;
+    subreq->super.input = src->input;
+    subreq->super.hostconf = src->hostconf;
+    subreq->super.pathconf = src->pathconf;
+    subreq->super.version = src->version;
+    subreq->super.error_logs = src->error_logs;
+    subreq->super.num_delegated = src->num_delegated;
+    subreq->super.num_reprocessed = src->num_reprocessed;
+    subreq->super.processed_at = src->processed_at;
+
+    return &subreq->super;
+}
+
+void h2o_dispose_subrequest(h2o_req_t *subreq)
```

Comment:
A function that allocates (or frees) the memory upon initialization should be named "foo_create" or "foo_destroy".
A function that receives a pointer to an object that needs to be initialized (or to be disposed of) should be named "foo_init" or "foo_dispose".